File: frontend/SunGreen-main/SunGreen-main/lib/screens/models/plant_details.dart
//models/plant_details.dart

class PlantDetails {
  final String plantId;
  final String name;
  final String type;
  final DateTime? lastWatered;
  final DateTime? nextWatering;
  final String? lastPhoto;
  final SensorData? lastSensorData;
  final String? lastRecommendation;

  PlantDetails({
    required this.plantId,
    required this.name,
    required this.type,
    this.lastWatered,
    this.nextWatering,
    this.lastPhoto,
    this.lastSensorData,
    this.lastRecommendation,
  });

  factory PlantDetails.fromJson(Map<String, dynamic> json) {
    return PlantDetails(
      plantId: json['plant_id'],
      name: json['name'],
      type: json['type'],
      lastWatered: json['last_watered'] != null && json['last_watered'] is String
          ? DateTime.tryParse(json['last_watered'])
          : null,
      nextWatering: json['next_watering'] != null && json['next_watering'] is String
          ? DateTime.tryParse(json['next_watering'])
          : null,
      lastPhoto: json['last_photo'],
      lastSensorData: json['last_sensor_data'] != null
          ? SensorData.fromJson(json['last_sensor_data'])
          : null,
      lastRecommendation: json['last_gemini_recommendation']?['text'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'plant_id': plantId,
      'name': name,
      'type': type,
      'last_watered': lastWatered?.toIso8601String().split("T")[0],
      'next_watering': nextWatering?.toIso8601String().split("T")[0],
      'last_photo': lastPhoto,
      'last_sensor_data': lastSensorData?.toJson(),
      'last_recommendation': lastRecommendation,
    };
  }

  String? get formattedLastWatered => lastWatered != null
      ? "${lastWatered!.year.toString().padLeft(4, '0')}-${lastWatered!.month.toString().padLeft(2, '0')}-${lastWatered!.day.toString().padLeft(2, '0')}"
      : null;

  String? get formattedNextWatering => nextWatering != null
      ? "${nextWatering!.year.toString().padLeft(4, '0')}-${nextWatering!.month.toString().padLeft(2, '0')}-${nextWatering!.day.toString().padLeft(2, '0')}"
      : null;
}

class SensorData {
  final double? temperature;
  final double? humidity;
  final double? soilMoisture;
  final double? light;
  final double? gasQuality;
  final DateTime? createdAt;

  SensorData({
    this.temperature,
    this.humidity,
    this.soilMoisture,
    this.light,
    this.gasQuality,
    this.createdAt,
  });

  factory SensorData.fromJson(Map<String, dynamic> json) {
    return SensorData(
      temperature: json['temperature']?.toDouble(),
      humidity: json['humidity']?.toDouble(),
      soilMoisture: json['soil_moisture']?.toDouble(),
      light: json['light']?.toDouble(),
      gasQuality: json['gas_quality']?.toDouble(),
      createdAt: json['created_at'] != null && json['created_at'] is String
          ? DateTime.tryParse(json['created_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'temperature': temperature,
      'humidity': humidity,
      'soil_moisture': soilMoisture,
      'light': light,
      'gas_quality': gasQuality,
      'created_at': createdAt?.toIso8601String(),
    };
  }
}

File: backend/routes/plant_details.py
#routes/plant_details.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from uuid import UUID
from models.database import get_db
from models.plant import Plant
from models.photo import Photo
from models.sensor_data import SensorData
from models.recommendation import Recommendation

router = APIRouter()


@router.get("/{plant_id}/details")
def get_plant_details(plant_id: UUID, db: Session = Depends(get_db)):
    plant = db.query(Plant).filter(Plant.id == plant_id).first()
    if not plant:
        raise HTTPException(status_code=404, detail="Plant not found")

    photo = (
        db.query(Photo)
        .filter(Photo.plant_id == plant_id)
        .order_by(Photo.created_at.desc())
        .first()
    )

    sensor = (
        db.query(SensorData)
        .filter(SensorData.plant_id == plant_id)
        .order_by(SensorData.created_at.desc())
        .first()
    )

    recommendation = (
        db.query(Recommendation)
        .filter(Recommendation.plant_id == plant_id)
        .order_by(Recommendation.created_at.desc())
        .first()
    )

    return {
    "plant_id": str(plant.id),
    "name": plant.name,
    "type": plant.type,
    "last_watered": plant.last_watered,
    "next_watering": recommendation.next_watering,  
    "last_photo": photo.s3_url if photo else None,
    "last_sensor_data": {
        "temperature": sensor.temperature,
        "humidity": sensor.humidity,
        "soil_moisture": sensor.soil_moisture,
        "light": sensor.light,
        "gas_quality": sensor.gas_quality,
        "created_at": sensor.created_at,
    } if sensor else {
        "temperature": None,
        "humidity": None,
        "soil_moisture": None,
        "light": None,
        "gas_quality": None,
        "created_at": None,
    },
    "last_gemini_recommendation": {
        "recommendation": recommendation.content,
        "next_watering_date": recommendation.next_watering,
        "created_at": recommendation.created_at,
    } if recommendation else {
        "recommendation": None,
        "next_watering_date": None,
        "created_at": None,
    }
}





